flow/CompressionUtils.cpp

```diff
@@ -81,27 +66,17 @@ StringRef CompressionUtils::compress(const CompressionFilter filter, const Strin
 	if (filter == CompressionFilter::NONE) {
 		return StringRef(arena, data);
 	}
-
-	namespace bio = boost::iostreams;
-	std::stringstream compStream;
-	std::stringstream decomStream(data.toString());
-
-	bio::filtering_streambuf<bio::input> out;
-#ifdef ZLIB_LIB_SUPPORTED
-	if (filter == CompressionFilter::GZIP) {
-		out.push(bio::gzip_compressor(bio::gzip_params(level)));
-	}
-#endif
 #ifdef ZSTD_LIB_SUPPORTED
 	if (filter == CompressionFilter::ZSTD) {
-		out.push(bio::zstd_compressor(bio::zstd_params(level)));
+		const char* src = reinterpret_cast<const char*>(data.begin());
+		size_t destSize = ZSTD_compressBound(data.size());
+		std::unique_ptr<uint8_t[]> dest = std::make_unique<uint8_t[]>(destSize);
```

Comment:
Alternate could be:

```
StringRef dst = makeString(destSize);
size_t bytes = ZSTD_compress(mutateString(dst), destSize, src, data.size(), level);
ASSERT(!ZSTD_isError(bytes));
return dst;
```
I think it would avoids an extra memcpy done when returning StringRef?

Reply:
ZSTD_compressBound doesn't return the exact size of compressed string. It's a "bound". The final "bytes" may be less than the bound.  If we return the dst buffer directly, it'll include garbage data

Reply:
In general for a CPU bound operations, the easiest optimization would be avoid memcpy's when possible; especially when looking from an operation end-to-end perspective. I buy the simplicity story, but, I guess above proposed code is simplistic enough and is already been used in Encyrption (so not that risky). 

Again, if you feel having an extra memcpy if negligible, it could be a judgement that you make :).

Reply:
Thank you for your input, Ata. Let's leave this topic open. I would do more experiements after zstd is enabled with prod load and see how much we can gain by eliminating the memcpy. For sure I would love to do it if it could be more helpful.

Reply:
Thanks Hui for the consideration. 
I think it would be better to optimize overall memcpy from end-to-end operation perspective, benchmark runs are useful but it doesn't always reflect the full end-to-end performance characteristics; also, to me it seems like a low-hanging optimization to do.

I might try to make the necessary change soon if possible.